File: 14a.py
#!/usr/bin/python3
import re


with open('14.dat') as f:
    data = f.read()
lines = data.split('\n')
lines.pop()
secs = 2503
regex = r"(\w+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds."
rein_data = list()
for line in lines:
    m = re.match(regex, line)
    rein_data.append((m.group(1), int(m.group(2)), int(m.group(3)), int(m.group(4))))
reindist = list()
for r in rein_data:
    (name, speed, runsecs, restsecs) = r
    reinruns = secs // (runsecs + restsecs)
    if (secs - reinruns * (runsecs + restsecs)) > runsecs:
        remain = runsecs
    else:
        remain = secs - reinruns * (runsecs + restsecs)
    reindist.append((reinruns * runsecs + remain) * speed)

maxdist = 0
maxrein = ''
for i, el in enumerate(rein_data):
    # print(el[0]+' '+str(reindist[i]))
    if maxdist < reindist[i]:
        maxdist = reindist[i]
        maxrein = el[0]
print(maxrein+' '+str(maxdist))

File: 14a.cc
#include <fstream>
#include <iostream>
#include <regex>
#include <string>

int main(void)
{
    std::string line;
    std::ifstream fin;
    std::smatch sm;
    std::vector<std::tuple<std::string, int, int, int>> rein_data;
    std::vector<int> rein_dist;
    std::regex re("(\\w+) can fly (\\d+) km/s for (\\d+) seconds, but then must rest for (\\d+) seconds.");
    int secs = 2503;

    fin.open("14.dat");
    std::getline(fin, line);
    while (!fin.eof())
    {
        if (std::regex_match(line, sm, re))
        {
            rein_data.push_back(std::make_tuple(sm[1], std::stoi(sm[2]), std::stoi(sm[3]), std::stoi(sm[4])));
        }
        std::getline(fin, line);
    }

    for (auto r : rein_data)
    {
        int remain;
        auto [name, speed, runsecs, restsecs] = r;
        int rein_runs = secs / (runsecs + restsecs);
        if ((secs - rein_runs * (runsecs + restsecs)) > runsecs)
        {
            remain = runsecs;
        }
        else
        {
            remain = secs - rein_runs * (runsecs + restsecs);
        }

        rein_dist.push_back((rein_runs * runsecs + remain) * speed);
    }

    int maxdist = 0;
    int i = 0;
    std::string maxrein = "";
    for (auto el : rein_data)
    {
        if (maxdist < rein_dist[i])
        {
            maxdist = rein_dist[i];
            maxrein = std::get<0>(el);
        }
        i++;
    }

    std::cout << maxrein << ' ' << maxdist << std::endl;

    fin.close();
}
